services/server/src/server/services/storageServices/AllianceDatabaseService.ts

```diff
@@ -0,0 +1,380 @@
+import {
+  Match,
+  CheckedContract,
+  Transformation,
+} from "@ethereum-sourcify/lib-sourcify";
+import { logger } from "../../../common/loggerLoki";
+import { id as keccak256str } from "ethers";
+import { IStorageService } from "../StorageService";
+import * as AllianceDatabase from "../utils/alliance-database-util";
+import { AuthTypes, Connector } from "@google-cloud/cloud-sql-connector";
+import { Pool } from "pg";
+
+export interface AllianceDatabaseServiceOptions {
+  googleCloudSql: {
+    instanceName: string;
+    iamAccount: string;
+    database: string;
+  };
+  postgres: {
+    host: string;
+    port: number;
+    database: string;
+    user: string;
+    password: string;
+  };
+}
+
+export class AllianceDatabaseService implements IStorageService {
+  private allianceDatabasePool?: any;
+
+  googleCloudSqlInstanceName?: string;
+  googleCloudSqlIamAccount?: string;
+  googleCloudSqlDatabase?: string;
+  postgresHost?: string;
+  postgresPort?: number;
+  postgresDatabase?: string;
+  postgresUser?: string;
+  postgresPassword?: string;
+
+  constructor(options: AllianceDatabaseServiceOptions) {
+    this.googleCloudSqlInstanceName = options.googleCloudSql.instanceName;
+    this.googleCloudSqlIamAccount = options.googleCloudSql.iamAccount;
+    this.googleCloudSqlDatabase = options.googleCloudSql.database;
+    this.postgresHost = options.postgres.host;
+    this.postgresPort = options.postgres.port;
+    this.postgresDatabase = options.postgres.database;
+    this.postgresUser = options.postgres.user;
+    this.postgresPassword = options.postgres.password;
+  }
+
+  async init(): Promise<boolean> {
+    // if the database is already initialized
+    if (this.allianceDatabasePool != undefined) {
+      return true;
+    }
+
+    try {
+      if (this.googleCloudSqlInstanceName) {
+        const connector = new Connector();
+        const clientOpts = await connector.getOptions({
+          instanceConnectionName: this.googleCloudSqlInstanceName, // "verifier-alliance:europe-west3:test-verifier-alliance",
+          authType: AuthTypes.IAM,
+        });
+        this.allianceDatabasePool = new Pool({
+          ...clientOpts,
+          user: this.googleCloudSqlIamAccount, // "marco.castignoli@ethereum.org",
+          database: this.googleCloudSqlDatabase, // "postgres",
+          max: 5,
+        });
+      } else if (this.postgresHost) {
+        this.allianceDatabasePool = new Pool({
+          host: this.postgresHost,
+          port: this.postgresPort,
+          database: this.postgresDatabase,
+          user: this.postgresUser,
+          password: this.postgresPassword,
+          max: 5,
+        });
+      } else {
+        throw new Error("Cannot initialize Alliance Database pool");
+      }
+      return true;
+    } catch (e) {
+      logger.error(e);
+      return false;
+    }
+  }
+
+  async storeMatch(recompiledContract: CheckedContract, match: Match) {
```

Comment:
This function is a little too long and complex. It would be nice if we can break it down into more understandable parts. I can't immediately think of a way. If not it's also ok. I know this is the workhorse of the Alliance service.

Reply:
I agree, but I think it's not related to this PR